### PR TITLE
Instructor: add courses/sessions: give more specific error messages when a field is empty #4129

### LINKED ISSUE #4129: Instructor: add courses/sessions: give more specific error messages when a field is empty
For example, instead of giving this as the error message,
"" is not acceptable to TEAMMATES as feedback session name because it is empty. The value of feedback session name should be no longer than 38 characters. It should not be empty

We can be direct and say the field cannot be empty.

### PR DESCRIPTION
**Pull request by amarlearning**
Fixes #4129 

Change error message from 

_[this]_

`
"" is not acceptable to TEAMMATES as feedback session name because it is empty. 
`

_[to this]_

`
The field <field_name> is empty. 
`

### CODE REVIEW

**Comment by amarlearning** [comment]
ready for review

File: src/main/java/teammates/common/util/FieldValidator.java [hunk 1]
```diff
@@ -111,6 +111,8 @@
     public static final String REASON_UNAVAILABLE_AS_CHOICE = "not available as a choice";
 
     // error message components
+    public static final String EMPTY_STRING_ERROR_INFO =
+            "The field ${fieldName} ${reason}.";
```
Comment (leeyimin) [inline src/main/java/teammates/common/util/FieldValidator.java]: `REASON_EMPTY` can be added here directly.

**Comment by amarlearning** [comment]
Ready for review

File: src/main/java/teammates/common/util/FieldValidator.java [hunk 2]
```diff
@@ -282,7 +292,7 @@ public String getInvalidityInfoForEmail(String email) {
         String sanitizedValue = SanitizationHelper.sanitizeForHtml(email);
 
         if (email.isEmpty()) {
-            return getPopulatedErrorMessage(EMAIL_ERROR_MESSAGE, email, EMAIL_FIELD_NAME, REASON_EMPTY,
+            return getPopulatedErrorMessage(EMAIL_ERROR_MESSAGE_EMPTY_STRING, email, EMAIL_FIELD_NAME, REASON_EMPTY,
```
Comment (leeyimin) [inline src/main/java/teammates/common/util/FieldValidator.java]: Create a new method to populate fields for empty string error messages.

File: src/test/java/teammates/test/cases/action/AdminEmailComposeSaveActionTest.java [hunk 3]
```diff
@@ -268,7 +268,7 @@ public void testExecuteAndPostProcess() throws Exception {
         expectedLogSegment = Const.ACTION_RESULT_FAILURE;
         AssertHelper.assertContains(expectedLogSegment, action.getLogMessage());
 
-        expectedStatus = "\"\" is not acceptable to TEAMMATES as a/an email subject";
+        expectedStatus = "The field email subject";
```
Comment (leeyimin) [inline src/test/java/teammates/test/cases/action/AdminEmailComposeSaveActionTest.java]: Put in a more complete version of the expected status.

File: src/test/java/teammates/test/cases/browsertests/InstructorFeedbackSessionsPageUiTest.java [hunk 4]
```diff
@@ -964,7 +964,8 @@ private void testValidationReload() throws Exception {
                 newSession.getStartTime(), newSession.getEndTime(), null, null,
                 newSession.getInstructions(),
                 newSession.getGracePeriod());
-        feedbackPage.verifyStatus(getPopulatedErrorMessage(FieldValidator.SIZE_CAPPED_NON_EMPTY_STRING_ERROR_MESSAGE,
+        feedbackPage.verifyStatus(getPopulatedErrorMessage(
+                                                    FieldValidator.SIZE_CAPPED_NON_EMPTY_STRING_ERROR_MESSAGE_EMPTY_STRING,
                                                            "", FieldValidator.FEEDBACK_SESSION_NAME_FIELD_NAME,
                                                            FieldValidator.REASON_EMPTY,
                                                            FieldValidator.FEEDBACK_SESSION_NAME_MAX_LENGTH));
```
Comment (leeyimin) [inline src/test/java/teammates/test/cases/browsertests/InstructorFeedbackSessionsPageUiTest.java]: The indentation of these lines can be improved.

File: src/main/java/teammates/common/util/FieldValidator.java (hunk 2 above)
Comment (amarlearning) [inline src/main/java/teammates/common/util/FieldValidator.java]: @leeyimin I don't understand why there is a need to create a new method just for `EMPTY_STRING`. I think we should revert [this](https://github.com/TEAMMATES/teammates/pull/7783#discussion_r127907532) change and use `getPopulatedErrorMessage` only.

Comment (leeyimin) [inline src/main/java/teammates/common/util/FieldValidator.java]: @amarlearning I believe since that message has been created just for empty string errors, there is little reason why it is necessary for users of that string to fill up the reason manually or fill in redundant fields using existing methods. Of course, what I suggested may not be the best way to improve your code, but I do think there is room for improvement.

**Comment by amarlearning** [comment]
Ready for review

File: src/main/java/teammates/common/util/FieldValidator.java [hunk 5]
```diff
@@ -751,4 +761,16 @@ private static String getPopulatedErrorMessage(
...
                               .replace("${reason}", errorReason);
     }
+
+    public static String getPopulatedEmptyStringErrorMessage(
+            String messageTemplate, String userInput, String fieldName, int maxLength) {
+        return getPopulatedEmptyStringErrorMessage(messageTemplate, userInput, fieldName)
+                   .replace("${maxLength}", String.valueOf(maxLength));
+    }
+
+    private static String getPopulatedEmptyStringErrorMessage(
```
Comment (leeyimin) [inline src/main/java/teammates/common/util/FieldValidator.java]: I believe you can inline this method.

File: src/main/java/teammates/common/util/FieldValidator.java [hunk 6]
```diff
@@ -751,4 +761,16 @@ private static String getPopulatedErrorMessage(
                               .replace("${fieldName}", fieldName)
                               .replace("${reason}", errorReason);
     }
+
+    public static String getPopulatedEmptyStringErrorMessage(
+            String messageTemplate, String userInput, String fieldName, int maxLength) {
```
Comment (leeyimin) [inline src/main/java/teammates/common/util/FieldValidator.java]: `userInput` is not a necessary field, I believe.

File: src/test/java/teammates/test/cases/BaseTestCase.java [hunk 7]
```diff
@@ -113,6 +113,20 @@ protected static String getPopulatedErrorMessage(String messageTemplate, String
                                      null, new Object[] { messageTemplate, userInput, fieldName, errorReason, maxLength });
     }
 
+    protected static String getPopulatedEmptyStringErrorMessage(String messageTemplate, String userInput,
```
Comment (leeyimin) [inline src/test/java/teammates/test/cases/BaseTestCase.java]: This method is not used.

File: src/test/java/teammates/test/cases/BaseTestCase.java [hunk 8]
```diff
@@ -113,6 +113,20 @@ protected static String getPopulatedErrorMessage(String messageTemplate, String
...
+                                                     String fieldName)
+            throws ReflectiveOperationException {
+        return getPopulatedEmptyStringErrorMessage(messageTemplate, userInput, fieldName, 0);
+    }
+
+    protected static String getPopulatedEmptyStringErrorMessage(String messageTemplate, String userInput,
+                                                     String fieldName, int maxLength)
+            throws ReflectiveOperationException {
+        return (String) invokeMethod(FieldValidator.class, "getPopulatedEmptyStringErrorMessage",
```
Comment (leeyimin) [inline src/test/java/teammates/test/cases/BaseTestCase.java]: Why not just use the method directly?

File: src/test/java/teammates/test/cases/action/AdminEmailComposeSaveActionTest.java [hunk 9]
```diff
@@ -268,7 +268,7 @@ public void testExecuteAndPostProcess() throws Exception {
         expectedLogSegment = Const.ACTION_RESULT_FAILURE;
         AssertHelper.assertContains(expectedLogSegment, action.getLogMessage());
 
-        expectedStatus = "\"\" is not acceptable to TEAMMATES as a/an email subject";
+        expectedStatus = "The field email subject is empty. The value of a/an email subject should be no";
```
Comment (leeyimin) [inline src/test/java/teammates/test/cases/action/AdminEmailComposeSaveActionTest.java]: Complete the sentence or leave it as "The field email subject is empty.".

File: src/main/java/teammates/common/util/FieldValidator.java (hunk 5 above)
Comment (amarlearning) [inline src/main/java/teammates/common/util/FieldValidator.java]: optimization :+1:

File: src/test/java/teammates/test/cases/action/AdminEmailComposeSaveActionTest.java (hunk 9 above)
Comment (amarlearning) [inline src/test/java/teammates/test/cases/action/AdminEmailComposeSaveActionTest.java]: Done :+1:

File: src/test/java/teammates/test/cases/BaseTestCase.java (hunk 7 above)
Comment (amarlearning) [inline src/test/java/teammates/test/cases/BaseTestCase.java]: :sweat_smile:

File: src/test/java/teammates/test/cases/BaseTestCase.java (hunk 8 above)
Comment (amarlearning) [inline src/test/java/teammates/test/cases/BaseTestCase.java]: @leeyimin I didn't get you. Can you explain me in brief what change are you expecting?

File: src/test/java/teammates/test/cases/BaseTestCase.java [hunk 10]
```diff
@@ -113,6 +113,13 @@ protected static String getPopulatedErrorMessage(String messageTemplate, String
                                      null, new Object[] { messageTemplate, userInput, fieldName, errorReason, maxLength });
     }
 
+    protected static String getPopulatedEmptyStringErrorMessage(String messageTemplate, String fieldName, int maxLength)
+            throws ReflectiveOperationException {
+        return (String) invokeMethod(FieldValidator.class, "getPopulatedEmptyStringErrorMessage",
```
Comment (leeyimin) [inline src/test/java/teammates/test/cases/BaseTestCase.java]: Can call the method directly, instead of using reflection.
`return FieldValidator.getPopulatedEmptyStringErrorMessage(...);`

**Comment by amarlearning** [comment]
Ready for review

File: src/test/java/teammates/test/cases/action/InstructorFeedbackEditCopyActionTest.java [hunk 11]
```diff
@@ -268,16 +268,16 @@ public void testExecuteAndPostProcess() {
 
         assertEquals("", editCopyData.redirectUrl);
 
-        expectedString = "\"\" is not acceptable to TEAMMATES as a/an feedback session name because it is empty. "
+        expectedString = "The field feedback session name is empty. "
```
Comment (damithc) [inline src/test/java/teammates/test/cases/action/InstructorFeedbackEditCopyActionTest.java]: On a related note, it is better if we can set the field name apart from the rest of the text. e.g. `The field 'feedback session name' is empty`. When it runs together with the rest of the test, it becomes hard to read. What do you think? Can be a separate PR.

Comment (amarlearning) [inline src/test/java/teammates/test/cases/action/InstructorFeedbackEditCopyActionTest.java]: @damithc It won't take more than a commit if you say I can do that in this PR only.

Comment (damithc) [inline src/test/java/teammates/test/cases/action/InstructorFeedbackEditCopyActionTest.java]: This issue might apply to, not just cases where the field is empty. Does it? If it is, then it goes beyond the scope of this PR?.

Comment (amarlearning) [inline src/test/java/teammates/test/cases/action/InstructorFeedbackEditCopyActionTest.java]: I think the issue is applicable only where the field is empty because for other errors we are still using this [`"Invalid name" is not acceptable to TEAMMATES as a/an feedback session name because it contains invalid character."`] as an error statement and I think we don't need to highlight the field in this case.

Comment (damithc) [inline src/test/java/teammates/test/cases/action/InstructorFeedbackEditCopyActionTest.java]: In that case we can do it in this PR. Go ahead.

**Comment by amarlearning** [comment]
Ready for review
